### ISSUE TITLE
Problem with heartbeats

### ISSUE THREAD

**Issue by kraih** [inway/mojo-rabbitmq-client#11]
This fairly simple script is supposed to run forever and print all the messages it receives.
```perl
use Mojo::Base -strict;
use Mojo::RabbitMQ::Client;

my $client = Mojo::RabbitMQ::Client->new(url => 'amqp://...:5672');
$client->on(error => sub { die pop });
$client->on(
  open => sub {
    my $client = shift;
    say 'Connected';
    my $channel = Mojo::RabbitMQ::Client::Channel->new();
    $channel->on(error => sub { die pop });
    $channel->on(
      open => sub {
        my $channel = shift;
        say 'Channel open';

        # Exchange
        my $exchange = $channel->declare_exchange(
          exchange => 'pubsub',
          type     => 'topic',
          passive  => 1,
          durable  => 1
        );
        $exchange->on(
          success => sub {
            say 'Declared exchange';

            # Declare queue
            my $queue = $channel->declare_queue(exclusive => 1);
            $queue->on(
              success => sub {
                my $result = pop;
                say 'Declared queue';

                # Bind queue
                my $queue_name = $result->method_frame->{queue};
                my $queue      = $channel->bind_queue(
                  exchange    => 'pubsub',
                  queue       => $queue_name,
                  routing_key => '#'
                );
                $queue->on(
                  success => sub {
                    say 'Bound queue';

                    # Consume messages
                    my $consumer
                      = $channel->consume(queue => $queue_name, no_ack => 1);
                    $consumer->on(
                      message => sub {
                        my $message = pop->{body}->to_raw_payload;
                        say $message;
                      }
                    );
                    $consumer->deliver;
                  }
                );
                $queue->deliver;
              }
            );
            $queue->deliver;
          }
        );
        $exchange->deliver;
      }
    );
    $client->open_channel($channel);
  }
);
$client->connect;

Mojo::IOLoop->start;
```
Which it does for a few seconds, then it produces this.
```
Mojo::Reactor::Poll: Timer failed: Can't call method "isa" on an undefined value at /usr/lib/perl5/vendor_perl/5.26.1/Mojo/RabbitMQ/Client.pm line 276.
```
And a few seconds later.
```
Mojo::Reactor::Poll: Timer failed: Consumer callback failure: Received data is not body frame at test.pl line 13.
Mojo::Reactor::Poll: Timer failed: Consumer callback failure: Received data is not header frame at test.pl line 13.
Mojo::Reactor::Poll: Timer failed: Consumer callback failure: Received data is not body frame at test.pl line 13.
Mojo::Reactor::Poll: Timer failed: Consumer callback failure: Received data is not header frame at test.pl line 13.
...
```
I was using Mojolicious 7.58 and Mojo::RabbitMQ::Client 0.0.9 for this experiment. Is there something obvious i was doing wrong, or have i stumbled over a bug?

**Comment by kraih** [comment]
It would be very helpful if you showed some information about the AMQP frames with an env var. Like we do in Mojolicious with MOJO_WEBSOCKET_DEBUG.

**Comment by SPodjasek** [comment]
That's strange. I have bunch of workers based on this module (but created using `Mojo::RabbitMQ::Client->consumer()` running on RabbitMQ 3.6.5 without any problems.

Could you please share your RabbitMQ server version?

**Comment by kraih** [comment]
I do not control the server and do not know the version. Could be 3.6.14.

**Comment by SPodjasek** [comment]
I'll try to do some checks tomorrow, but it looks like an empty frame was received and not properly handled by code.

**Comment by kraih** [comment]
Please let me know if there is anything i can do to help.

**Comment by SPodjasek** [comment]
Just tested your code with RabbitMQ 3.5.7, Erlang 18.3, Mojolicious 6.66, Net::AMQP 0.06, Perl v5.23.6 and works like a charm. I just had to drop `passive => 1` for exchange to declare properly.
So probably the issue lies somewhere in dependency tree.

**Comment by kraih** [comment]
If i remove the `passive => 1` it never reaches the success or error events and nothing happens.

**Comment by kraih** [comment]
Btw. That version of Mojolicious is fairly [out of date](https://github.com/kraih/mojo/blob/3f9f53845f1d7c0b0892929bdfbb01af0100adab/Changes#L373).

**Comment by kraih** [comment]
I just got confirmation that the server i'm testing against also runs RabbitMQ 3.5.7.

**Comment by SPodjasek** [comment]
I know, my perl build environment is pretty rusty right now. 
In the meantime of brewing Perl 5.27.7 to test everything on Mojolicious@master I've made  `Client` to use `$ENV{MOJO_RABBITMQ_DEBUG}` to enable listening of wire messages, and some initial method exchange in 801d367.

**Comment by SPodjasek** [comment]
You can see your working example wire log here: https://gist.github.com/SPodjasek/481d67cb42f105bef05665fe50bebbde

**Comment by kraih** [comment]
That looks very useful.

**Comment by kraih** [comment]
So, i've started doing some tests again, and this problem still appears to exist. What i'm seeing with `MOJO_RABBITMQ_DEBUG` activated is very strange. It appears that things break when there is multiple AMQP frames in the same read operation. (I've removed most actual message data since i couldn't make sure there is nothing confidential in there, but kept all the protocol metadata)
```
<- "\1\0\1\0\0\0S\0<\0<\37amq.ctag-eC9dqmBs8xIlgD4nOpcdyA\0\0\0\0\0\0\0\333\0\6pubsub\36suse.obs.package.build_success\316\2\0\1\0\0\0)\0<\0\0\0\0\0\0\0\0\1\201\230\0\30application/octet-stream\2\0\316\3\0\1\0\0\1\201{\"project\":...}\316\1\0\1\0\0\0S\0<\0<\37amq.ctag-eC9dqmBs8xIlgD4nOpcdyA\0\0\0\0\0\0\0\334\0\6pubsub\36suse.obs.package.build_success\316\2\0\1\0\0\0)\0<\0\0\0\0\0\0\0\0\1\200\230\0\30application/octet-stream\2\0\316\3\0\1\0\0\1\200{\"project\":...}\316\1\0\1\0\0\0U\0<\0<\37amq.ctag-eC9dqmBs8xIlgD4nOpcdyA\0\0\0\0\0\0\0\335\0\6pubsub suse.obs.package.build_unchanged\316\2\0\1\0\0\0)\0<\0\0\0\0\0\0\0\0\1w\230\0\30application/octet-stream\2\0\316\3\0\1\0\0\1w{\"project\":..."repository\":\"standar"
Mojo::Reactor::Poll: Timer failed: Can't call method "isa" on an undefined value at /usr/lib/perl5/vendor_perl/5.26.2/Mojo/RabbitMQ/Client.pmline 318.
```
Yes, the frame is cut off like that at the end. I've been testing it with this consumer on a moderately active exchange (maybe 0.5-2 messages per second on average).
```perl
use Mojo::Base -strict;

use Mojo::Promise;
use Mojo::RabbitMQ::Client;

Mojo::IOLoop->singleton->reactor->unsubscribe('error');

my $client = Mojo::RabbitMQ::Client->new(url => 'amqp://...:5672');
my $queue_name;
$client->connect_p->then(sub { shift->acquire_channel_p() })->then(sub {
  shift->declare_exchange_p(
    exchange => 'pubsub',
    type     => 'topic',
    passive  => 1,
    durable  => 1
  );
})->then(sub { shift->declare_queue_p(exclusive => 1) })->then(sub {
  my ($channel, $result) = @_;
  $queue_name = $result->method_frame->{queue};
  $channel->bind_queue_p(
    exchange    => 'pubsub',
    queue       => $queue_name,
    routing_key => '#'
  );
})->then(sub {
  my $channel = shift;

  my $promise = Mojo::Promise->new;
  my $consumer = $channel->consume(queue => $queue_name, no_ack => 1);
  $consumer->on(error => sub { $promise->reject('Consumer error') });
  $consumer->on(
    message => sub {
      my $frame   = pop;
      my $key     = $frame->{deliver}->method_frame->routing_key;
      my $message = $frame->{body}->to_raw_payload;
      say "\n*** $key: " . length $message;
    }
  );
  $consumer->deliver;

  return $promise;
})->catch(sub { warn shift })->wait;
```
Hope this helps with tracking down the problem.

**Comment by kraih** [comment]
Looking at the responsible code, there appears to be something very very wrong. `Net::AMQP->parse_raw_frames` is already removing the frames it parses from the buffer, yet there is custom buffer manipulation happening [here](https://metacpan.org/source/SEBAPOD/Mojo-RabbitMQ-Client-0.1.0/lib/Mojo/RabbitMQ/Client.pm#L314). If i understand `Net::AMQP` correctly you shouldn't have to mess with the protocol, ever, and just let it deal with the low level bits.

**Comment by kraih** [comment]
The actual bug here seems to be that `_parse_frames` recurses with `next_tick` and then removes half a frame from the buffer, which causes the error. The correct fix will be to get rid of all that useless buffer handling. I'll see if i can fix it tomorrow.

**Comment by kraih** [comment]
I think this problem has been resolved with the 0.2.0 release.
